File: go/vec_gen_opt.go
// Optimizations suggested by kjk, see http://blog.cdleary.com/2012/06/simple-selfish-and-unscientific-shootout/#comment-546669602

package main

import (
    "bufio"
    "fmt"
    "log"
    "math/rand"
    "os"
    "strconv"
)

func main() {
    if len(os.Args) != 2 {
        fmt.Fprintf(os.Stderr, "usage: %s <elem_count>\n", os.Args[0])
        os.Exit(-1)
    }

    count, err := strconv.Atoi(os.Args[1])
    if err != nil { panic(err) }

    file, err := os.Create("vec_gen.out")
    if err != nil { panic(err) }
    defer file.Close()

    bw := bufio.NewWriterSize(file, 1024*16)
    defer func() {
        err := bw.Flush()
        if err != nil { log.Fatal(err) }
    }()

    var scratch [128]byte
    for i := 0; i < count; i++ {
        b := strconv.AppendUint(scratch[:0], uint64(rand.Uint32()), 10)
        b2 := append(b, 10)
        if _, err := bw.Write(b2); err != nil { panic(err) }
    }
}

File: lua/vec_gen_opt.lua
-- Written by Mike Pall, see http://blog.cdleary.com/2012/06/simple-selfish-and-unscientific-shootout/#comment-546907272

local ffi = require("ffi")
ffi.cdef[[
typedef struct FILE FILE;
int fprintf(FILE *, const char *, ...);
]]
if #arg ~= 1 then
io.stderr:write("Usage: " .. arg[0] .. " <elem_count>\n")
os.exit(-1)
end    
local count = tonumber(arg[1])
math.randomseed(os.time())
local fp = io.open("vec_gen.out", "w")
for i = 1,count do
ffi.C.fprintf(fp, "%d\n", ffi.new("int", math.random(0, 2^31-1)))
end
fp:close()
